### DISCUSSION TITLE
Would I be able to make this project compatible with Lightning Returns?

### DISCUSSION THREAD

**Discussion by cvlnomen** [rebtd7/FF13Fix#76] [Q&A]
Today I found out that I can use OneTweakNG to reduce the amount of stutter that occur in Lightning Returns while loading new areas. This fix does not work anymore with the changes made by this project. When [comparing v1.6.4 with v0.2](https://github.com/rebtd7/FF13Fix/compare/0.2...rebtd7:1.6.4), it does not seem like too much has changed between the two projects. This made me think that it would be possible to make this project compatible with LR. My question is how hard you guys think this would be and if I would be able to do it? 

Some things about myself:
I am an IT student who is going to follow a College Minor about developing a game engine in C++. 
I already have 2 years of programming experience with C#, Dart and Golang. 
I am now learning the basics of C++. 
I almost don't have any experience with game development and modding.

**Comment by rebtd7** [comment]
Hmm... are you sure that OneTweakNG reduces the stuttering?
I don't think it does anything generic that could improve the LR's performance (except perhaps for forcing triple buffering, but this setting is disabled by default)
@Nucleoprotein any idea on what could be happening here?

I've never taken a look in the game so I could be mistaken, but checking the issues reported for the game I don't think there's anything FF13Fix added to OneTweakNG that could help LR (hacks are FF13 and FF13-2 specific). I don't think there's anything to port. If OneTweakNG is somehow improving performance, you should probably keep using that.

> My question is how hard you guys think this would be and if I would be able to do it?

I don't think fixing the issues in the game would be easy, specially considering you don't have experience with modding/reverse engineering. Expect to spend dozen/hundreds of hours creating breakpoints, monitoring RAM and reading/experimenting with ASM code.

That said, the only way to acquire experience is to try to do something 😄 

I come from the romhacking/fantranslation scene; I haven't done too much work in reverse engineering PC games. I personally think hacking PC games is harder: there are no save states to experiment, and the games execute too many instructions per second: it is harder to trace anything. X86 and X86_64 assembly code/call conventions is also messier compared to RISC architectures like ARM/MIPS code imo.
So don't get discouraged if you don't get anywhere with LR (perhaps try something on a weaker platform with a good debugger emulator).

In the PC scene it seems people use [Ghidra](https://ghidra-sre.org/) to do static analysis of code and to decompile ASM to C, but I don't really have experience with that (it is also not trivial depending on how the game executable is encrypted). 

FF13Fix was basically done checking the games' assembly code in real time using [Cheat Engine](https://www.cheatengine.org/), and using [PIX for Windows](https://www.microsoft.com/en-us/download/details.aspx?id=6812) (an old unsupported version included with the DX9SDK, as the new ones are not compatible with D3D9), and [APITrace](https://apitrace.github.io/) to try to understand its graphics API calls. Logging wrapped/hooked API calls with the DLL is also very useful.

**Comment by Nucleoprotein** [comment]
I will analyze this when I'm back from hospital (in about a week).

niedz., 18 lip 2021, 17:35 użytkownik rebtd7 ***@***.***>
napisał:

**Reply by rebtd7** [reply]
> I will analyze this when I'm back from hospital (in about a week).

hey, I hope you get better soon!
